### PR TITLE
Port to futures 0.3 and tokio 0.2.0

### PR DESCRIPTION
**Pull request by sdbondi**
Ported this library to tokio 0.2.0-alpha and to use async/await syntax. 

I didn't manage to upgrade the hyper example yet, but I added a quick example of usage with tor.

Hope this is of use to you :)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -16,16 +16,15 @@ edition = "2018"
 travis-ci = { repository = "sticnarf/tokio-socks" }
 
 [dependencies]
-futures = "0.1"
-tokio-tcp = "0.1"
-tokio-udp = "0.1"
-tokio-io = "0.1"
+futures= {version= "0.3.0-alpha.19", package="futures-preview"}
+tokio-net = {version="0.2.0-alpha.6", features=["tcp"]}
+tokio-io = {version="0.2.0-alpha.6", features=["util"]}
```
Comment (sticnarf) [inline Cargo.toml]: Could you make it well formatted? Like:
```toml
tokio-net = { version = "0.2.0-alpha.6", features = ["tcp"] }
```

File: examples/hyper.rs [hunk 2]
```diff
@@ -1,52 +1,54 @@
...
-use futures::prelude::*;
-use hyper::{
-    client::connect::{Connect, Connected, Destination},
-    Client, Uri,
-};
-use std::io::{prelude::*, stdout};
-use std::net::SocketAddr;
-use tokio_socks::{tcp::Socks5Stream, Error};
-use tokio_tcp::TcpStream;
+//use failure::{Compat, Fail};
```
Comment (sticnarf) [inline examples/hyper.rs]: I think you can just remove this file for now. Hyper examples can be added later.

File: src/error.rs [hunk 3]
```diff
@@ -71,5 +71,5 @@ impl From<std::io::Error> for Error {
     }
 }
 
-/// Result type of `tokio-socks`
-pub type Result<T> = std::result::Result<T, Error>;
+///// Result type of `tokio-socks`
+//pub type Result<T> = std::result::Result<T, Error>;
```
Comment (sticnarf) [inline src/error.rs]: I prefer to keep the `Result` type. Then, we can omit the `Error` in other code.

File: src/lib.rs [hunk 4]
```diff
@@ -1,37 +1,39 @@
+//#![feature(type_alias_impl_trait)]
```
Comment (sticnarf) [inline src/lib.rs]: Just remove it?

File: src/tcp.rs [hunk 5]
```diff
@@ -1,19 +1,28 @@
...
 use std::net::{Ipv4Addr, Ipv6Addr, SocketAddr};
-use tokio_io::{AsyncRead, AsyncWrite};
-use tokio_tcp::{ConnectFuture as TokioConnect, TcpStream};
+use std::pin::Pin;
+use tokio_io::{AsyncReadExt, AsyncWriteExt, AsyncWrite};
+use tokio_net::tcp::TcpStream;
+use futures::stream::Fuse;
+use tokio_io::AsyncRead;
+use std::io;
+use std::io::ErrorKind;
```
Comment (sticnarf) [inline src/tcp.rs]: Could you combine some of these imports and make them in order?

File: src/tcp.rs [hunk 6]
```diff
@@ -113,37 +154,49 @@ impl Socks5Stream {
...
     ptr: usize,
     len: usize,
 }
 
-impl<'a, 't, S> ConnectFuture<'a, 't, S>
-where
-    S: Stream<Item = SocketAddr, Error = Error>,
+impl<'a, 't, S> SocksConnector<'a, 't, S>
+    where
+        S: Stream<Item=Result<SocketAddr, Error>> + Unpin,
```
Comment (sticnarf) [inline src/tcp.rs]: The code seems not formatted.

File: src/tcp.rs [hunk 7]
```diff
@@ -1,19 +1,28 @@
...
 
 #[repr(u8)]
 #[derive(Clone, Copy)]
 enum Command {
     Connect = 0x01,
     Bind = 0x02,
+    #[allow(dead_code)]
     Associate = 0x03,
+    TorResolve = 0xF0,
+    TorResolvePtr = 0xF1,
```
Comment (sticnarf) [inline src/tcp.rs]: Tor related features are not standard socks5. Could you make them under a feature flag?

**Comment by sticnarf** [comment]
Version in `Cargo.toml` can be increase to `0.2`

File: src/tcp.rs (hunk 7 above)
Comment (sdbondi) [reply src/tcp.rs]: Good idea!

File: src/tcp.rs (hunk 6 above)
Comment (sdbondi) [reply src/tcp.rs]: Will fix :)

File: src/tcp.rs (hunk 5 above)
Comment (sdbondi) [reply src/tcp.rs]: Sure thing :)

File: src/lib.rs (hunk 4 above)
Comment (sdbondi) [reply src/lib.rs]: Will do - forgot to remove it

File: src/error.rs (hunk 3 above)
Comment (sdbondi) [reply src/error.rs]: No problem, think there were a few places where I'd have to use `std::result::Result` but I don't mind either way

File: examples/hyper.rs (hunk 2 above)
Comment (sdbondi) [reply examples/hyper.rs]: Perfect - thanks

File: Cargo.toml (hunk 1 above)
Comment (sdbondi) [reply Cargo.toml]: Wlll do

File: src/tcp.rs (hunk 5 above)
Comment (sdbondi) [reply src/tcp.rs]: I added a rustfmt.toml which makes rustfmt do this for you

**Comment by sdbondi** [comment]
Thanks for the review - I've made the changes you suggested 🎉

**Comment by sdbondi** [comment]
@sticnarf Any chance this could be merged and released?

**Comment by sticnarf** [comment]
The dependencies need to be upgraded to the latest version. Also in the end I hope I can add the hyper example back.

**Comment by sticnarf** [comment]
@sdbondi  Thanks very much!

**Pull request merged**